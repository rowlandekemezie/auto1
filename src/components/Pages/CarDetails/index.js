import React, { useEffect } from "react";
import styled from "styled-components";
import Placeholder from "react-placeholder";
import PropTypes from "prop-types";
import { connect } from "react-redux";

import NavBar from "../../NavBar";
import Footer from "../../Footer";
import Button from "../../Button";

import { carActions } from "../../../redux/actions";
import { capitalize } from "../../../utils";

const Main = styled("main")`
  flex: 1;
  display: flex;
  flex-direction: column;
  padding: 12px;
  max-width: 800px;
  margin: 0 auto;
  justify-content: space-between;
  min-height: 100vh;
  min-height: calc(100vh - 184px);

  @media (min-width: 768px) {
    flex-direction: row;
    padding: 24px;
  }
`;

const CallToActionBox = styled("div")`
  border: 2px solid ${({ theme }) => theme.GRAY};
  transition: all 0.3s ease;
  padding: 12px;
  margin-bottom: 12px;
  height: 30%;
  display: flex;
  flex-direction: column;
  padding: 24px;

  &:focus,
  &:hover {
    box-shadow: 5px 5px 25px 0px rgba(46, 61, 73, 0.2);
  }

  @media (min-width: 768px) {
    width: 30%;
  }
`;

const Description = styled("div")`
  justify-content: center;
  light-height: 18px;

  @media (min-width: 768px) {
    width: 50%;
  }
`;

const Text = styled("p")`
  font-size: 14px;
  line-height: 24px;
  text-align: justify;
`;

const Title = styled("p")`
  font-weight: 700;
  font-size: 32px;
  margin-top: 8px;
`;

const SaveButtonWrapper = styled("div")`
  align-self: flex-end;
`;

const SubTitle = styled("p")`
  font-size: 18px;
  font-weight: 500;
`;

function DescriptionCard({ car }) {
  return (
    <Description>
      <Title>{car.modelName}</Title>
      <SubTitle>
        Stock #{car.stockNumber} - {car.mileage.number} KM - {car.fuelType} -{" "}
        {capitalize(car.color)}{" "}
      </SubTitle>
      <Text>
        This is currently available and can be delivered as soon as tomorrow
        morning. Please be aware that delivery times shown in their page are not
        definitive and may change due to bad weather conditions.
      </Text>
    </Description>
  );
}

export function CarDetails({
  fetchACar,
  carId,
  selectedCar,
  saveCollection,
  collection
}) {
  const handleOnSave = () => saveCollection(selectedCar);
  const isSaved = collection.find(i => i.stockNumber === parseInt(carId));

  useEffect(() => {
    fetchACar(carId);
  }, [carId]);

  return (
    <Placeholder ready={!!selectedCar} type="media" rows={7} delay={200}>
      <NavBar />
      <Main>
        <DescriptionCard car={selectedCar} />
        <CallToActionBox>
          <Text>
            If you like this car, click the button and save it in your
            collection fo favourite items.
          </Text>
          <SaveButtonWrapper>
            <Button
              label={`${isSaved ? "Remove from collection" : "Save"}`}
              onClick={handleOnSave}
            />
          </SaveButtonWrapper>
        </CallToActionBox>
      </Main>
      <Footer />
    </Placeholder>
  );
}

const mapStateToProps = (
  { cars: { selectedCar, isLoading, collection } },
  ownParams
) => ({
  selectedCar,
  isLoading,
  collection,
  carId: ownParams.match.params.carId
});

const mapDispatchToProps = {
  fetchACar: carActions.fetchACar,
  saveCollection: carActions.saveCollection
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(CarDetails);

CarDetails.propTypes = {
  car: PropTypes.shape({
    selectedCar: PropTypes.shape({
      modelName: PropTypes.string.isRequired,
      stockNumber: PropTypes.number.isRequired,
      color: PropTypes.string.isRequired,
      fuelType: PropTypes.string.isRequired,
      mileage: PropTypes.shape({
        number: PropTypes.number.isRequired
      })
    }),
    isLoading: PropTypes.bool
  }),
  carId: PropTypes.string.isRequired,
  fetchACar: PropTypes.func.isRequired,
  collection: PropTypes.array
};
