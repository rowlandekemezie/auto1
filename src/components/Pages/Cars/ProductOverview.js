import React from "react";
import styled from "styled-components";
import PlaceHolder from "react-placeholder";
import PropTypes from "prop-types";
import Link from "react-router-dom/Link";

import FilterBox from "../../FilterBox";
import Pagination from "../../Pagination";

import { capitalize } from "../../../utils";

const Wrapper = styled("div")`
  display: flex;
  flex-direction: column;
  width: 100%;

  @media (min-width: 768px) {
    width: 66.666%;
  }
`;

const CardWrapper = styled("div")`
  display: flex;
  border: 2px solid ${({ theme }) => theme.GRAY};
  transition: all 0.3s ease;
  padding: 12px;
  margin-bottom: 12px;

  &:focus,
  &:hover {
    box-shadow: 5px 5px 25px 0px rgba(46, 61, 73, 0.2);
  }
`;

const Description = styled("div")`
  justify-content: center;
  light-height: 18px;
`;

const StyledLink = styled(Link)`
  color: ${({ theme }) => theme.ORANGE};
  font-size: 12px;
  text-decoration: none;
`;

const Image = styled("img")`
  width: 120px;
  height: 120px;
  margin-right: 24px;
`;

const Text = styled("p")`
  font-size: 12px;
  font-weight: 400;
`;

const Title = styled("p")`
  font-weight: 700;
  font-size: 18px;
  margin-top: 8px;
`;

const SortFilterBoxWrapper = styled("div")`
  margin-bottom: 24px;

  @media (min-width: 768px) {
    width: 40%;
  }
`;

const HeadingContainer = styled("div")`
  display: flex;
  flex-direction: column;
  width: 100%;

  @media (min-width: 768px) {
    flex-direction: row;
    justify-content: space-between;
  }
`;

const PageMetaData = styled("div")`
  width: 60%;
`;

const SubTitle = styled("p")`
  font-size: 18px;
`;

function Card({ car }) {
  return (
    <CardWrapper>
      <Image src={car.pictureUrl} />
      <Description>
        <Title>{car.modelName}</Title>
        <Text>
          Stock #{car.stockNumber} - {car.mileage.number} KM - {car.fuelType} -{" "}
          {capitalize(car.color)}{" "}
        </Text>
        <StyledLink to={`/cars/${car.stockNumber}`}>View details</StyledLink>
      </Description>
    </CardWrapper>
  );
}

const SORTING_OPTIONS = [
  { value: "", label: "None" },
  { value: "asc", label: "Mileage - Ascending Order" },
  { value: "des", label: "Mileage - Descending Order" }
];

export default function ProductOverview({
  cars,
  value,
  onNext,
  isLoading,
  onPrevious,
  currentPage,
  previousPage,
  totalPageCount,
  handleOnChange
}) {
  return (
    <PlaceHolder ready={!!cars || isLoading} type="media" row={7}>
      <Wrapper>
        <HeadingContainer>
          <PageMetaData>
            <Title>Available cars</Title>
            <SubTitle>
              Showing {cars.length} of {totalPageCount}{" "}
              {`result${cars.length > 1 ? "s" : ""}`}{" "}
            </SubTitle>
          </PageMetaData>
          <SortFilterBoxWrapper>
            <FilterBox
              value={value}
              options={SORTING_OPTIONS}
              title="Sort by"
              handleOnChange={handleOnChange}
            />
          </SortFilterBoxWrapper>
        </HeadingContainer>
        {cars.map((car, index) => (
          <Card car={car} key={`${car.stockNumber}-${index}`} />
        ))}
        <Pagination
          currentPage={currentPage}
          previousPage={previousPage}
          totalPageCount={totalPageCount}
          onNext={onNext}
          onPrevious={onPrevious}
        />
      </Wrapper>
    </PlaceHolder>
  );
}

ProductOverview.propTypes = {
  cars: PropTypes.arrayOf(
    PropTypes.shape({
      modelName: PropTypes.string.isRequired,
      stockNumber: PropTypes.number.isRequired,
      color: PropTypes.string.isRequired,
      fuelType: PropTypes.string.isRequired,
      mileage: PropTypes.shape({
        number: PropTypes.number.isRequired
      })
    }).isRequired
  ),
  onNext: PropTypes.func.isRequired,
  onPrevious: PropTypes.func.isRequired,
  handleOnChange: PropTypes.func.isRequired,
  totalPageCount: PropTypes.number,
  value: PropTypes.object,
  currentPage: PropTypes.number,
  previousPage: PropTypes.number
};
