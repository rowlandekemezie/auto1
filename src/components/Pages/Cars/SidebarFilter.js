import React, { useState, useEffect } from "react";
import styled from "styled-components";
import PropTypes from "prop-types";

import Button from "../../Button";
import FilterBox from "../../FilterBox";

import { renderOptions, buildQueryString } from "../../../utils";

const Wrapper = styled("div")`
  display: flex;
  flex-direction: column;
  margin-right: 24px;
  padding: 24px;
  border: 2px solid ${({ theme }) => theme.GRAY};
  height: 30%;
  margin-right: 0;

  @media (min-width: 768px) {
    width: 33.333%;
    margin-right: 24px;
  }
`;

const FilterButtonWrapper = styled("div")`
  align-self: flex-end;
  margin-top: 24px;
`;

export default function SidebarFilter({
  fetchAllColors,
  fetchAllManufacturers,
  manufacturers: { manufacturers },
  colors: { colors },
  history,
  fetchAllCars
}) {
  const [color, setColor] = useState({ value: "", label: "All car colors" });
  const [manufacturer, setManufacturer] = useState({
    value: "",
    label: "All manufacturers"
  });

  useEffect(() => {
    fetchAllManufacturers();
    fetchAllColors();
  }, []);

  const handleFilter = () => {
    const filters = {
      manufacturer: manufacturer.value,
      color: color.value,
      page: 1
    };
    const query = buildQueryString(filters);
    history.push(`/?${query}`);
    fetchAllCars(query);
  };

  return (
    <Wrapper>
      <FilterBox
        value={color}
        handleOnChange={e => setColor(e)}
        title="color"
        options={renderOptions(colors, "All car colors")}
      />
      <FilterBox
        value={manufacturer}
        handleOnChange={e => setManufacturer(e)}
        title="manufacturer"
        options={renderOptions(manufacturers, "All manufacturers")}
      />
      <FilterButtonWrapper>
        <Button label="Filter" onClick={handleFilter} />
      </FilterButtonWrapper>
    </Wrapper>
  );
}

SidebarFilter.propTypes = {
  fetchAllColors: PropTypes.func.isRequired,
  fetchAllManufacturers: PropTypes.func.isRequired,
  manufacturers: PropTypes.shape({
    manufacturers: PropTypes.arrayOf(PropTypes.string.isRequired)
  }).isRequired,
  colors: PropTypes.shape({
    colors: PropTypes.arrayOf(PropTypes.string.isRequired)
  }).isRequired,
  history: PropTypes.shape({
    push: PropTypes.func.isRequired
  }),
  fetchAllCars: PropTypes.func.isRequired
};
