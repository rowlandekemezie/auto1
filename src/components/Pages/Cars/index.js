import React, { useEffect, useState } from "react";
import styled from "styled-components";
import Placeholder from "react-placeholder";
import PropTypes from "prop-types";
import { connect } from "react-redux";

import Navbar from "../../Navbar";
import Footer from "../../Footer";
import ProductOverview from "./ProductOverview";
import SidebarFilter from "./SidebarFilter";

import {
  carActions,
  colorActions,
  manufacturerActions
} from "../../../redux/actions";
import { buildQueryString, initialPageNumber } from "../../../utils";

const Main = styled("main")`
  flex: 1;
  display: flex;
  flex-direction: column;
  padding: 12px;
  justify-content: space-between;
  min-height: calc(100vh - 184px);

  @media (min-width: 768px) {
    flex-direction: row;
    padding: 24px;
  }
`;

const ErrorMessage = styled("p")`
  text-align: center;
  color: ${({ theme }) => theme.ORANGE};
`;

export function Cars({
  cars,
  error,
  colors,
  history,
  isLoading,
  fetchAllCars,
  manufacturers,
  totalPageCount,
  fetchAllColors,
  fetchAllManufacturers
}) {
  const [page, setCurrentPage] = useState(initialPageNumber());
  const [previousPage, setPreviousPage] = useState(page);
  const [sort, setSort] = useState({ value: "", label: "None" });

  useEffect(() => {
    const filters = { page, sort: sort.value };
    const query = buildQueryString(filters);
    fetchAllCars(query);

    history.push(`/?${query}`);
  }, [page, sort]);

  const onNext = pageNum => {
    setPreviousPage(page);
    setCurrentPage(pageNum);
  };

  return (
    <Placeholder ready={!!cars} type="media" rows={7}>
      <Navbar />
      <Main>
        <SidebarFilter
          fetchAllColors={fetchAllColors}
          fetchAllManufacturers={fetchAllManufacturers}
          manufacturers={manufacturers}
          colors={colors}
          history={history}
          fetchAllCars={fetchAllCars}
        />
        <ProductOverview
          cars={cars}
          value={sort}
          currentPage={page}
          previousPage={previousPage}
          onNext={onNext}
          onPrevious={setCurrentPage}
          totalPageCount={totalPageCount}
          handleOnChange={e => setSort(e)}
        />
        {error && <ErrorMessage>{error}</ErrorMessage>}
      </Main>
      <Footer />
    </Placeholder>
  );
}

const mapStateToProps = (
  { cars: { cars, isLoading, totalPageCount }, manufacturers, colors, error },
  ownParams
) => ({
  cars,
  isLoading,
  totalPageCount,
  colors,
  manufacturers
});

const mapDispatchToProps = {
  fetchAllCars: carActions.fetchAllCars,
  fetchAllColors: colorActions.fetchAllColors,
  fetchAllManufacturers: manufacturerActions.fetchAllManufacturers
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Cars);

Cars.propTypes = {
  cars: PropTypes.arrayOf(
    PropTypes.shape({
      modelName: PropTypes.string.isRequired,
      stockNumber: PropTypes.number.isRequired,
      color: PropTypes.string.isRequired,
      fuelType: PropTypes.string.isRequired,
      mileage: PropTypes.shape({
        number: PropTypes.number.isRequired
      })
    })
  ),
  isLoading: PropTypes.bool,
  totalPageCount: PropTypes.number,
  fetchAllColors: PropTypes.func.isRequired,
  fetchAllManufacturers: PropTypes.func.isRequired,
  manufacturers: PropTypes.shape({
    manufacturers: PropTypes.arrayOf(PropTypes.string.isRequired)
  }).isRequired,
  colors: PropTypes.shape({
    colors: PropTypes.arrayOf(PropTypes.string.isRequired)
  }).isRequired,
  history: PropTypes.shape({
    push: PropTypes.func.isRequired
  }),
  fetchAllCars: PropTypes.func.isRequired
};
