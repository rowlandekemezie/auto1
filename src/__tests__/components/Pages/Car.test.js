import React from "react";
import { shallow, mount } from "enzyme";
import { Cars } from "../../../components/Pages/Cars";

describe("<Cars />", () => {
  const props = {
    fetchAllColors: jest.fn(),
    fetchAllManufacturers: jest.fn(),
    manufacturers: { manufacturers: ["BMW"] },
    colors: { colors: ["Red", "Yellow"] },
    fetchAllCars: jest.fn(),
    history: { push: jest.fn() }
  };

  it("should render component properly", () => {
    const wrapper = shallow(<Cars {...props} />);
    expect(wrapper.length).toBeTruthy();
    expect(wrapper).toMatchSnapshot();
  });

  it("should render with the required proptypes", () => {
    const wrapper = mount(<Cars {...props} />);
    expect(wrapper.props()).toBeObject(props);
    expect(wrapper.props().fetchAllCars).toBeFunction();
    expect(wrapper.props().fetchAllManufacturers).toBeFunction();
    expect(wrapper.props().fetchAllColors).toBeFunction();
    expect(wrapper.props().manufacturers).toBeObject(props.manufacturers);
    expect(wrapper.props().colors).toBeObject(props.colors);
  });
});
