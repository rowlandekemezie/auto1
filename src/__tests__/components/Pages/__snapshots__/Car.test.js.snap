// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Cars /> should render component properly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Cars
    colors={
      Object {
        "colors": Array [
          "Red",
          "Yellow",
        ],
      }
    }
    fetchAllCars={[MockFunction]}
    fetchAllColors={[MockFunction]}
    fetchAllManufacturers={[MockFunction]}
    history={
      Object {
        "push": [MockFunction],
      }
    }
    manufacturers={
      Object {
        "manufacturers": Array [
          "BMW",
        ],
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <NavBar />,
        <ForwardRef>
          <SidebarFilter
            colors={
              Object {
                "colors": Array [
                  "Red",
                  "Yellow",
                ],
              }
            }
            fetchAllCars={[MockFunction]}
            fetchAllColors={[MockFunction]}
            fetchAllManufacturers={[MockFunction]}
            history={
              Object {
                "push": [MockFunction],
              }
            }
            manufacturers={
              Object {
                "manufacturers": Array [
                  "BMW",
                ],
              }
            }
          />
          <ProductOverview
            currentPage={1}
            handleOnChange={[Function]}
            onNext={[Function]}
            onPrevious={[Function]}
            previousPage={1}
            value={
              Object {
                "label": "None",
                "value": "",
              }
            }
          />
        </ForwardRef>,
        <Footer />,
      ],
      "color": "#CDCDCD",
      "delay": 0,
      "ready": false,
      "rows": 7,
      "type": "media",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": Array [
            <SidebarFilter
              colors={
                Object {
                  "colors": Array [
                    "Red",
                    "Yellow",
                  ],
                }
              }
              fetchAllCars={[MockFunction]}
              fetchAllColors={[MockFunction]}
              fetchAllManufacturers={[MockFunction]}
              history={
                Object {
                  "push": [MockFunction],
                }
              }
              manufacturers={
                Object {
                  "manufacturers": Array [
                    "BMW",
                  ],
                }
              }
            />,
            <ProductOverview
              currentPage={1}
              handleOnChange={[Function]}
              onNext={[Function]}
              onPrevious={[Function]}
              previousPage={1}
              value={
                Object {
                  "label": "None",
                  "value": "",
                }
              }
            />,
            undefined,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "colors": Object {
                "colors": Array [
                  "Red",
                  "Yellow",
                ],
              },
              "fetchAllCars": [MockFunction],
              "fetchAllColors": [MockFunction],
              "fetchAllManufacturers": [MockFunction],
              "history": Object {
                "push": [MockFunction],
              },
              "manufacturers": Object {
                "manufacturers": Array [
                  "BMW",
                ],
              },
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "cars": undefined,
              "currentPage": 1,
              "handleOnChange": [Function],
              "onNext": [Function],
              "onPrevious": [Function],
              "previousPage": 1,
              "totalPageCount": undefined,
              "value": Object {
                "label": "None",
                "value": "",
              },
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          undefined,
        ],
        "type": Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "componentStyle": ComponentStyle {
            "componentId": "sc-ckVGcZ",
            "isStatic": true,
            "rules": Array [
              "
  flex: 1;
  display: flex;
  flex-direction: column;
  padding: 12px;
  justify-content: space-between;
  min-height: calc(100vh - 184px);

  @media (min-width: 768px) {
    flex-direction: row;
    padding: 24px;
  }
",
            ],
          },
          "displayName": "styled.main",
          "foldedComponentIds": Array [],
          "render": [Function],
          "styledComponentId": "sc-ckVGcZ",
          "target": "main",
          "toString": [Function],
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        },
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <NavBar />,
          <ForwardRef>
            <SidebarFilter
              colors={
                Object {
                  "colors": Array [
                    "Red",
                    "Yellow",
                  ],
                }
              }
              fetchAllCars={[MockFunction]}
              fetchAllColors={[MockFunction]}
              fetchAllManufacturers={[MockFunction]}
              history={
                Object {
                  "push": [MockFunction],
                }
              }
              manufacturers={
                Object {
                  "manufacturers": Array [
                    "BMW",
                  ],
                }
              }
            />
            <ProductOverview
              currentPage={1}
              handleOnChange={[Function]}
              onNext={[Function]}
              onPrevious={[Function]}
              previousPage={1}
              value={
                Object {
                  "label": "None",
                  "value": "",
                }
              }
            />
          </ForwardRef>,
          <Footer />,
        ],
        "color": "#CDCDCD",
        "delay": 0,
        "ready": false,
        "rows": 7,
        "type": "media",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": Array [
              <SidebarFilter
                colors={
                  Object {
                    "colors": Array [
                      "Red",
                      "Yellow",
                    ],
                  }
                }
                fetchAllCars={[MockFunction]}
                fetchAllColors={[MockFunction]}
                fetchAllManufacturers={[MockFunction]}
                history={
                  Object {
                    "push": [MockFunction],
                  }
                }
                manufacturers={
                  Object {
                    "manufacturers": Array [
                      "BMW",
                    ],
                  }
                }
              />,
              <ProductOverview
                currentPage={1}
                handleOnChange={[Function]}
                onNext={[Function]}
                onPrevious={[Function]}
                previousPage={1}
                value={
                  Object {
                    "label": "None",
                    "value": "",
                  }
                }
              />,
              undefined,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "colors": Object {
                  "colors": Array [
                    "Red",
                    "Yellow",
                  ],
                },
                "fetchAllCars": [MockFunction],
                "fetchAllColors": [MockFunction],
                "fetchAllManufacturers": [MockFunction],
                "history": Object {
                  "push": [MockFunction],
                },
                "manufacturers": Object {
                  "manufacturers": Array [
                    "BMW",
                  ],
                },
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "cars": undefined,
                "currentPage": 1,
                "handleOnChange": [Function],
                "onNext": [Function],
                "onPrevious": [Function],
                "previousPage": 1,
                "totalPageCount": undefined,
                "value": Object {
                  "label": "None",
                  "value": "",
                },
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            undefined,
          ],
          "type": Object {
            "$$typeof": Symbol(react.forward_ref),
            "attrs": Array [],
            "componentStyle": ComponentStyle {
              "componentId": "sc-ckVGcZ",
              "isStatic": true,
              "rules": Array [
                "
  flex: 1;
  display: flex;
  flex-direction: column;
  padding: 12px;
  justify-content: space-between;
  min-height: calc(100vh - 184px);

  @media (min-width: 768px) {
    flex-direction: row;
    padding: 24px;
  }
",
              ],
            },
            "displayName": "styled.main",
            "foldedComponentIds": Array [],
            "render": [Function],
            "styledComponentId": "sc-ckVGcZ",
            "target": "main",
            "toString": [Function],
            "warnTooManyClasses": [Function],
            "withComponent": [Function],
          },
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
